R/room.R

```diff
@@ -0,0 +1,101 @@
+#' Create the representation of a matrix room.
+#'
+#' @param id Matrix API identifier of the room.
+#' @param since Date and time of the beginning of the included events.
+#' @param events A `data.frame` containing the room events.
+#' @param next_token Token for the next sync. This should be set from the
+#'   `next_batch` item from the last `sync()` that was used to update this
+#'   room.
+#'
+#' @return An object of class `ChatStat_room` that stores the parameters as
+#'   named list items.
+#'
+#' @noRd
+room <- function(id,
+                 since = lubridate::now(),
+                 events = tibble::tibble(),
+                 next_token = NULL) {
+  stopifnot(is.character(id) & length(id) == 1)
+
+  structure(
+    list(
+      id = id,
+      since = since,
+      events = events,
+      next_token = next_token
+    ),
+    class = "ChatStat_room"
+  )
+}
+
+#' Update the room state using the Matrix API.
+#'
+#' This function extends the existing state of the room with new events from
+#' the Matrix API. Events are added from a recent call to [sync()] and the gap
+#' to the existing timeline is filled with additional calls to the messages
+#' API.
+#'
+#' @param room The room to update.
+#' @param sync The result of a recent call to [sync()]. This will be called
+#'   automatically, if the parameter is missing. However, it is recommended to
+#'   share the sync between multiple rooms to avoid duplicated work and to have
+#'   a common base for comparing the timelines.
+#'
+#' @export
+update_room <- function(room, sync = NULL) {
+  if (!inherits(room, "ChatStat_room")) {
+    stop(
+      "Room can't be updated, because it is not a valid object of class ",
+      "\"ChatStat_room\". See ChatStat::get_room() for creating one."
+    )
+  }
+
+  if (is.null(room$next_token)) {
+    stop(
+      "Room can't be updated, because it has not been initialized using the ",
+      "API. Use ChatStat::get_room() to get an updatable room object."
+    )
+  }
+
+  if (is.null(sync)) {
+    sync <- sync(since = room$next_token)
+  }
+
+  events <- room$events
+  next_batch_token <- room$next_token
+  prev_batch_token <- sync$rooms$join[[room$id]]$timeline$prev_batch
```

Comment:
If I understand this right, it means we need to sync every time we call update rooms, yes? I tried:

```
s <- sync()
r <- get_room(id, date, s)
update_room(r, s)
```

And the new events were not retrieved. If I call `update_room(r)` instead it works as expected, so that implies an old sync cannot be re-used - which isn't really how Matrix is supposed to work - you sync once and then go from there.

Assuming I have that right, I may try to debug it ... :)

Reply:
Well, maybe I'm misunderstanding the matrix spec here, but I thought that the `since` parameter of sync was exactly for that – to extend the results of a last sync. So you are supposed to do:

```r
initial_sync <- sync()
room <- get_room(…)

# Time passes…

new_sync <- sync(since = room$next_token)
room <- update_room(room, new_sync)
```